src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1234,25 +1380,30 @@ public static TSource Last<TSource>(this IEnumerable<TSource> source)
 
         public static TSource Last<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate)
         {
-            if (source == null) throw Error.ArgumentNull("source");
-            if (predicate == null) throw Error.ArgumentNull("predicate");
-            TSource result = default(TSource);
-            bool found = false;
-            foreach (TSource element in source)
+            Error.CheckSourceNotNull(source);
+            Error.CheckPredicateNotNull(predicate);
+            using (IEnumerator<TSource> e = source.GetEnumerator())
             {
-                if (predicate(element))
+                while (e.MoveNext())
                 {
-                    result = element;
-                    found = true;
+                    TSource result = e.Current;
+                    if (predicate(result))
+                    {
+                        while (e.MoveNext())
+                        {
+                            TSource element = e.Current;
+                            if (predicate(element)) result = element;
+                        }
```

Comment:
Understandable is in the eye of the beholder; I find this approach a better logical match to the task than the previous. It's probably the least beneficial change in this whole PR though, and largely included because it matches the style of a lot of the other changes.